stanford-earth/SE3_BLT
# Build test environment

# DO NOT MERGE

# User stories

- As a developer I want to run all of the behat tests for my site on my local environment.
- It would be nice to have the tests live in the module's repo as well.
- As a developer I want to always have the appropriate version of behat tests for each module in my project.

# Summary
This is a working demo of one solution to the problems listed above.  It would allow developers to run Behat tests with the BLT command, `blt tests:all` from within a controlled environment, in this case, the Docker version of DrupalVM.

Controlling the test environment is important because Behat relies on a number of packages and programs that may differ between development machines, for example, do you have Firefox and/or Chrome installed?  What versions?  And frequently, a new version that might automatically install on your machine, breaks Behat's ability to self-navigate websites.  Not to mention, Behat requires an agonizing amount of custom configuration.  When on-boarding a new developer, it would usually require 1-2 hours of support before the new, unique environment was ready to run tests.

Our hope is that tests will be able to travel with their associated module or site.  And that anyone from project managers to TravisCI will be able to run tests with minimal setup and specialized troubleshooting.

Why Behat, DrupalVM, Docker, BLT?
- **Behat**: There are many reasons we like this testing tool, not the least of which we have a great deal of experience and expertise on staff.  At the end of the day, whether a website "works" is our primary concern, and Behat tests at that level.
- **DrupalVM**: This decision rests on shakier grounds.  In forum discussions, it sounds like members of the Drupal community are using a wide variety of pre-packed containers and VM's to serve their testing needs.  Two things I like about DrupalVM include, I don't have to do any additional work like setup and install MySQL to get a Drupal site working.  It comes with sensible defaults.  And the fearless Jeff Geerling appears to be dedicated to documentation, for the most part, and very responsive to issues, bugs, etc. as they develop.  Finally, there appears to be some old integration between BLT and DrupalVM (the VirtualBox/Vagrant version).  Part of me hopes that work on integration between DrupalVM and BLT will continue into the Docker version.
- **Docker**: This is fairly easy, several members of our team have said the slowness of VirtualBox is a non-starter for them.  Plus, Docker images are smaller, which means we may be able to pull them onto a service like Travis, build a site, and run tests (fingers crossed; I have read about others doing something like this).
- **BLT**: We're already using BLT for a number of DevOps related tasks.  We discussed allowing tests to remain in various directories throughout the website's file structure.  BLT is extensible and being transitioned from Phing to Robo.  Robo appears to be the leading contender for a PHP task runner, so I am interested in seeing if we can build this logic as a BLT extension.

This is marked DO NOT MERGE because we'd like to abstract everything out into its own composer package.  And because I need to clean up some of the config files for BLT, DrupalVM, and Behat.

# Steps to test
You will need to have docker installed and running for this.  If on a Mac, Docker can be downloaded from: https://store.docker.com/editions/community/docker-ce-desktop-mac.  For other operating systems, visit Get Docker at: https://www.docker.com.

RUN A PRE-BUILT VERSION OF THE DOCKER CONTAINER (~5m)
1. Download https://stanford.box.com/s/3gin7u2vlkwgi4yq7os2e12q6wdde7p2 to ~/Sites
2. `cd ~/Sites/se3_blt/vendor/geerlingguy/drupal-vm`
3. Run `composer docker-load-image`
4. `../../../run-baked-container.sh`
5. Log into the container with `docker exec -it earth bash`
4. Once logged into the container, `cd /var/www/earth`
6. Then run, `blt local:setup`
9. `cd tests/behat`
10. And run `../../vendor/bin/behat -p default features`
11. This should run two simple behat test

BUILD THE DOCKER CONTAINER FROM SCRATCH (~15-25m)
1. Checkout the test-environment branch
2. Run `composer update`
3. Run `git status` and checkout vendor files that appear changed by composer update.
2. Run ./bake.sh
3. Wait (~10m) for this script to complete
4. Once logged into the container, `cd /var/www/earth`
6. Then run, `blt local:setup`
9. `cd tests/behat`
10. And run `../../vendor/bin/behat -p default features`
11. This should run two simple behat test

Notes:
- This currently installs the standard profile.  Even so, the install will fail to import uuid's.
- If you want to refresh the database from production, ie. behat local:refresh, you'll need to add your keys to ~/.ssh and a config file for passing your key to Acquia (ask if you want more details).
- Instead of step 10, I'd like to eventually use blt tests:behat, however that currently results in the error, "h: 1: lsof: not found," which is also being discussed here: https://github.com/acquia/blt/issues/1382.
- These configurations assumes you have nothing running on your machine's ports 9000 and 9001.

References:
https://blt.readthedocs.io/en/stable/readme/testing/#functional-testing
https://blt.readthedocs.io/en/stable/readme/local-development/#drupal-vm-and-behat-tests
